src/libcmd/misc-store-flags.hh

```diff
@@ -0,0 +1,21 @@
+#include "args.hh"
+#include "content-address.hh"
+
+namespace nix::flag {
+
+Args::Flag hashAlgo(std::string && longName, HashAlgorithm * ha);
+static inline Args::Flag hashAlgo(HashAlgorithm * ha)
+{
+    return hashAlgo("hash-algo", ha);
+}
+Args::Flag hashAlgoOpt(std::string && longName, std::optional<HashAlgorithm> * oha);
+Args::Flag hashFormatWithDefault(std::string && longName, HashFormat * hf);
+Args::Flag hashFormatOpt(std::string && longName, std::optional<HashFormat> * ohf);
+static inline Args::Flag hashAlgoOpt(std::optional<HashAlgorithm> * oha)
+{
+    return hashAlgoOpt("hash-algo", oha);
+}
+Args::Flag fileIngestionMethod(FileIngestionMethod * method);
+Args::Flag contentAddressMethod(ContentAddressMethod * method);
```

Comment:
Why do we have two nearly identical flags?

Reply:
They are different data types intentionally.

- [`FileContentAddress`](https://github.com/NixOS/nix/blob/d2c6a93bd5a213e66c5e6774776889eecbe3e625/src/libutil/file-content-address.hh#L11-L12) is the different ways we can content-address *file system objects* into a hash
- [`StoreContentAddress`](https://github.com/NixOS/nix/blob/d2c6a93bd5a213e66c5e6774776889eecbe3e625/src/libstore/content-address.hh#L42) is the different ways we can content-address *store objects* (i.e. references too) into a store path

FWIW, I was planning on writing some long-form documentation on this in the store chapter after this PR got the code in roughly the right shape.

Reply:
`nix hash path` I interpret as just working on the level of file system objects, as it doesn't deal with store path-based reference or other store concepts. `nix store add` however definitely creates store objects, and could indeed support arbitrary references. (Hell, it could even support trusted users unsafely creating input-addressed store objects too.)

Reply:
Ah right, and we didn't expose these to the command-line before since without `text` there's an isomorphism between these.

It feels a bit confusing and dangerous to have two different things here :thinking:

Reply:
Well, I think the main reason we didn't before was not any high-minded philosophical reason, but the fact that until recently `text` had entirely separate duplicative code paths.

Reply:
It doesn't feel confusing or dangerous to me; the CLI will match the internals and the docs I would next write, fostering a consistent mental model across all 3.

Reply:
We can also always backwards-compatibly have `nix hash path` support `text` later, so they once again match, and simply document that as far as file system objects are concerned `flat` and `text` are the same.

(I don't like this as much, as I think it makes superficial matters clearer and deep matters less clear, but for anyone that is unsure, it is good to know that this is kept around as an option. We can make sure that `FileIngestionMethod` (and its names) remains a sub-enum of `ContentAddressMethod` to preserve this option too.)